src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/WcfTestCase.cs

```diff
@@ -44,13 +50,48 @@ internal WcfTestCase(IXunitTestCase testCase, string skippedReason = null, bool
 
         public void Deserialize(IXunitSerializationInfo info) { _testCase.Deserialize(info); }
 
-        public Task<RunSummary> RunAsync(
+        public async Task<RunSummary> RunAsync(
             IMessageSink diagnosticMessageSink, IMessageBus messageBus, object[] constructorArguments,
             ExceptionAggregator aggregator, CancellationTokenSource cancellationTokenSource)
         {
-            return _isTheory
-                ? new XunitTheoryTestCaseRunner(this, DisplayName, _skippedReason, constructorArguments, _diagnosticMessageSink, messageBus, aggregator, cancellationTokenSource).RunAsync()
-                : new XunitTestCaseRunner(this, DisplayName, _skippedReason, constructorArguments, TestMethodArguments, messageBus, aggregator, cancellationTokenSource).RunAsync();
+            ConcurrentQueue<EventWrittenEventArgs> events = new ConcurrentQueue<EventWrittenEventArgs>();
+	          s_testListener._eventWritten = events.Enqueue;
+
+            RunSummary runsummary = await (_isTheory ? new XunitTheoryTestCaseRunner(this, DisplayName, _skippedReason, constructorArguments, _diagnosticMessageSink, messageBus, aggregator, cancellationTokenSource).RunAsync()
+                : new XunitTestCaseRunner(this, DisplayName, _skippedReason, constructorArguments, TestMethodArguments, messageBus, aggregator, cancellationTokenSource).RunAsync());
+
+            if (runsummary.Failed > 0)
+            {
+                s_testListener._eventWritten = null;
+                StringBuilder etwOutput = new StringBuilder();
+                etwOutput.AppendLine(string.Format("---ETW Trace for Test {0} Begins---", DisplayName));
+                foreach (var item in events)
+                {
+                    try
+                    {
+                        etwOutput.AppendLine(string.Format(DisplayName + ":" + item.Message, item.Payload));
```

Comment:
>string.Format(DisplayName + ":" + item.Message, item.Payload [](start = 45, length = 60)

Is this String.Format the way you intended?  I can't tell what it will do. It is passing DisplayName and item.Message as the format template -- which could easily blow up if they contain formatting characters like {1}. Was it intended to be something like String.Format("{0}:{1} {2}", DisplayName, item.Message, item.Payload)?